doc/source/data/dataset-execution-model.rst

```diff
@@ -0,0 +1,110 @@
+Memory Management
+-----------------
+
+This section deals with how Datasets manages execution and object store memory.
+
+Execution Memory
+~~~~~~~~~~~~~~~~
+
+During execution, certain types of intermediate data must fit in memory. This includes the input block of a task, as well as at least one of the output blocks of the task (when a task has multiple output blocks, only one needs to fit in memory at any given time). The input block consumes object stored shared memory (Python heap memory for non-Arrow data). The output blocks consume Python heap memory (prior to putting in the object store) as well as object store memory (after being put in the object store).
+
+This means that large block sizes can lead to potential out of memory situations. To avoid OOM errors, Datasets tries to split blocks during map and read tasks into pieces smaller than the target max block size. In some cases this splitting is not possible (i.e., if a single item in a block is extremely large, or the function given to ``.map_batches`` returns a very large batch). To avoid these issues, make sure no single item in your Datasets is too large, and always call ``.map_batches`` with batch size small enough that the output batch can comfortably fit into memory.
+
+Object Store Memory
+~~~~~~~~~~~~~~~~~~~
+
+Datasets uses the Ray object store to store data blocks, which means it inherits the memory management features of the Ray object store. This section discusses the relevant features:
+
+**Object Spilling**: Since Datasets uses the Ray object store to store data blocks, any blocks that can't fit into object store memory are automatically spilled to disk. The objects are automatically reloaded when needed by downstream compute tasks:
+
+.. image:: dataset-spill.svg
+   :width: 650px
+   :align: center
+
+..
+  https://docs.google.com/drawings/d/1H_vDiaXgyLU16rVHKqM3rEl0hYdttECXfxCj8YPrbks/edit
+
+**Locality Scheduling**: Ray will preferentially schedule compute tasks on nodes that already have a local copy of the object, reducing the need to transfer objects between nodes in the cluster.
+
+**Reference Counting**: Dataset blocks are kept alive by object store reference counting as long as there is any Dataset that references them. To free memory, delete any Python references to the Dataset object.
+
+**Load Balancing**: Datasets uses Ray scheduling hints to spread read tasks out across the cluster to balance memory usage.
+
+
+Performance Tips and Tuning
+---------------------------
+
+Batching Transforms
+~~~~~~~~~~~~~~~~~~~
+
+Mapping individual records using ``.map(fn)`` can be quite slow. Instead, consider using ``.map_batches(batch_fn, batch_format="pandas")`` and writing your ``batch_fn`` to perform vectorized pandas operations.
+
+Parquet Column Pruning
+~~~~~~~~~~~~~~~~~~~~~~
+
+Current Datasets will read all Parquet columns into memory. Make sure to specify the list of columns explicitly when calling ``ray.data.read_parquet()`` to avoid loading unnecessary data.
```

Comment:
```suggestion
Current Datasets will read all Parquet columns into memory. If you only need a subset of the columns, make sure to specify the list of columns explicitly when calling ``ray.data.read_parquet()`` to avoid loading unnecessary data.
```